File: src/Imaging/ComputeAngles.jl
"""
**ComputeAngles**

*Compute angles for shot gathers. These angles can be used for mapping migrated shots into angle gathers during shot profile migration.*

**IN**


* angx = "angx" : filename for incidence angles in the x direction for each shot
* angy = "angy" : filename for incidence angles in the y direction for each shot
* dip_flag = false : flag to subtract reflector dip from the computed angles to make them with reference to reflector normal
* vel = "vel" : seis file containing the velocity (should have same x and z dimensions as the desired image)
* wav = "wav" : seis file containing the source wavelet (in time domain)
* sz = 0. : source depth (Dev: read this from source wavelet file for variable source depth)
* nhx = 101 : number of offset bins
* ohx = 1000. : min offset (surface offset in the data)
* dhx = 10. : offset increment
* nhy = 101 : number of offset bins
* ohy = 1000. : min offset (surface offset in the data)
* dhy = 10. : offset increment
* pade_flag = false : flag for Pade Fourier correction
* fmin = 0. : min frequency to process (Hz)
* fmax = 80. : max frequency to process (Hz)
* padt = 2 : pad factor for the time axis
* padx = 2 : pad factor for the spatial axes
* verbose = false : flag for error / debugging messages
* sx = [0.] : array of source X positions (meters)
* sy = [0.] : array of source Y positions (meters)

**OUT**

*Credits: AS, 2015*

"""

function ComputeAngles(angx::AbstractString,angy::AbstractString,dip_flag=false,vel="vel",wav="wav",sz=0.,nhx=100,ohx=0,dhx=10,nhy=1,ohy=0,dhy=10,pade_flag=false,fmin=0,fmax=80,padt=2,padx=2,verbose=false,sx=[0],sy=[0])

	nshot = length(sx)
	if (dip_flag=="y")
		dipx,h = SeisRead(dipx_name)
		dipy,h = SeisRead(dipy_name)
	end

	v,h = SeisRead(vel)
	min_imx = h[1].imx
	max_imx = h[end].imx
	dmx = dhx
	nmx = max_imx - min_imx + 1
	min_imy = h[1].imy
	max_imy = h[end].imy
	dmy = dhy
	nmy = max_imy - min_imy + 1
	nz = h[1].n1
	dz = h[1].d1
	oz = h[1].o1
	min_gx = h[1].mx
	max_gx = h[end].mx
	min_gy = h[1].my
	max_gy = h[end].my

	shot_list = Array(ShotAngles,nshot)
	for ishot = 1 : nshot
		shot_list[ishot] = ShotAngles(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
		shot_list[ishot].angx = join(["angx_shot_" int(floor(sx[ishot])) "_" int(floor(sy[ishot]))])
		shot_list[ishot].angy = join(["angy_shot_" int(floor(sx[ishot])) "_" int(floor(sy[ishot]))])
		shot_list[ishot].vel = vel
		shot_list[ishot].dipx = dipx_name
		shot_list[ishot].dipy = dipy_name
		shot_list[ishot].wav = wav
		shot_list[ishot].sx = sx[ishot]
		shot_list[ishot].sy = sy[ishot]
		shot_list[ishot].sz = sz
		shot_list[ishot].ohx = ohx
		shot_list[ishot].dhx = dhx
		shot_list[ishot].nhx = nhx
		shot_list[ishot].ohy = ohy
		shot_list[ishot].dhy = dhy
		shot_list[ishot].nhy = nhy
		shot_list[ishot].fmin = fmin
		shot_list[ishot].fmax = fmax
		shot_list[ishot].omp = omp
		shot_list[ishot].pade_flag = pade_flag
		shot_list[ishot].dip_flag = (dip_flag == true) ? "y" : "n"
		shot_list[ishot].verbose = (verbose == true) ? "y" : "n"
	end

	a = pmap(compute_angles,shot_list)

	j = 1
	for ishot = 1 : nshot
		angx_shot,h_shot = SeisRead(shot_list[ishot].angx)
		angy_shot,h_shot = SeisRead(shot_list[ishot].angy)
		SeisWrite(angx,angx_shot,h_shot,itrace=j)
		SeisWrite(angy,angy_shot,h_shot,itrace=j)
		j += size(angx_shot,2)
		SeisRemove(shot_list[ishot].angx)
		SeisRemove(shot_list[ishot].angy)
	end

end

type ShotAngles
	angx
	angy
	vel
	dipx
	dipy
	wav
	sx
	sy
	sz
	ohx
	dhx
	nhx
	ohy
	dhy
	nhy
	fmin
	fmax
	pade_flag
	dip_flag
	omp
	verbose
end

function compute_angles(shot)
	ENV["OMP_NUM_THREADS"] = shot.omp
	if (find_library(["compute_angles"]) == "")
		error("Couldn't find shared library compute_angles.so, make sure it is installed and check LD_LIBRARY_PATH environmental variable.")
	end
	argv = ["0",
	join(["angx=",shot.angx]), join(["angy=",shot.angy]), join(["vel=",shot.vel]), join(["dipx=",shot.dipx]), join(["dipy=",shot.dipy]), join(["wav=",shot.wav]),
	join(["sx=",shot.sx]), join(["sy=",shot.sy]),  join(["sz=",shot.sz]),
	join(["ohx=",shot.ohx]),  join(["dhx=",shot.dhx]),  join(["nhx=",shot.nhx]),
	join(["ohy=",shot.ohy]),  join(["dhy=",shot.dhy]),  join(["nhy=",shot.nhy]),
	join(["fmin=",shot.fmin]),  join(["fmax=",shot.fmax]),
	join(["pade_flag=",shot.pade_flag]),  join(["dip_flag=",shot.dip_flag]),  join(["verbose=",shot.verbose]) ]
	a = ccall((:main, "compute_angles"), Int32, (Int32, Ptr{Ptr{UInt8}}), length(argv), argv)
	return(a)

end

File: src/Imaging/c/compute_angles.c
#include "seismic.h"
#include "wem.h"

void docs () {
	char * doc =
		"                                                                      \n"
		" Common shot wave equation angle of incidence computation for 3D acoustic, isotropic data. \n"
		"                                                                      \n";
	fprintf (stderr, "%s", doc);
}

int main (int argc, char *argv[])
{
	char angx_name[512],angy_name[512],vel_name[512],wav_name[512],dipx_name[512],dipy_name[512];
	struct SeisHeader *h_angx;
	struct SeisHeader *h_angy;
	struct SeisHeader *h_vel;
	struct SeisHeader *h_dipx,*h_dipy;
	struct SeisHeader *h_wav;
	int nx,ny,nz,nt,ix,iz,imx,imy;
	float **angx,**angy,**vel,**dipx,**dipy,**wav,sx,sy,sz;
	float ox,dx,oy,dy,oz,dz,ot,dt,fmin,fmax;
	int ntraces;
	bool verbose,dip_flag;
	struct SeisFileHeader fh;
	
	if (!par_read_bool(argc,argv,"verbose",&verbose)) verbose = false;
	if (!par_read_bool(argc,argv,"dip_flag",&dip_flag)) dip_flag = false; // flag to correct angles relative to reflector normal
	if (!par_read_string(argc,argv,"angx", angx_name)) { docs (); exit (1); }
	if (!par_read_string(argc,argv,"angy", angy_name)) { docs (); exit (1); }
	if (!par_read_string(argc,argv,"vel", vel_name)) { docs (); exit (1); }
	if (!par_read_string(argc,argv,"dipx", dipx_name)) { docs (); exit (1); }
	if (!par_read_string(argc,argv,"dipy", dipy_name)) { docs (); exit (1); }
	if (!par_read_string(argc,argv,"wav", wav_name)) { docs (); exit (1); }
	if (!par_read_float(argc,argv,"sz",&sz)) sz = 0;
	if (!par_read_float(argc,argv,"sx",&sx)) sx = 0;
	if (!par_read_float(argc,argv,"sy",&sy)) sy = 0;	
	if (!par_read_float(argc,argv,"fmin",&fmin)) fmin = 0;
	if (!par_read_float(argc,argv,"fmax",&fmax)) fmax = 80;
	// get dimensions from velocity (nz,oz,dz,nx,ox,dx) and wavelet (nt,sx) files
	InitFileHeader(&fh);
	ReadFileHeader(wav_name,&fh);
	nt = fh.n1;
	ReadFileHeader(vel_name,&fh);
	nz = fh.n1; ntraces = fh.n2*fh.n3*fh.n4*fh.n5;
	//fprintf(stderr,"nx=%d\n",nx);
	//fprintf(stderr,"nz=%d\n",nz);
	//fprintf(stderr,"nt=%d\n",nt);
	h_wav = allocSeisHeader(1);
	wav = alloc2float(nt,1); 		
	SeisRead(wav_name,wav,h_wav,&fh);
	ot = h_wav[0].o1;
	dt = h_wav[0].d1;
	h_vel = allocSeisHeader(ntraces);
	vel = alloc2float(nz,ntraces);
	SeisRead(vel_name,vel,h_vel,&fh); 	
	nx = h_vel[ntraces-1].imx - h_vel[0].imx + 1;
	ny = h_vel[ntraces-1].imy - h_vel[0].imy + 1;
	oz = h_vel[0].o1;
	dz = h_vel[0].d1;
	ox = h_vel[0].mx;
	dx = h_vel[1].mx - h_vel[0].mx;
	oy = h_vel[0].my;
	dy = ntraces > nx ? h_vel[nx].my - h_vel[nx-1].my : dx;
	angx = alloc2float(nz,nx*ny);
	angy = alloc2float(nz,nx*ny);
	for (iz=0;iz<nz;iz++) for (ix=0;ix<nx*ny;ix++) angx[ix][iz] = 0.0;
	for (iz=0;iz<nz;iz++) for (ix=0;ix<nx*ny;ix++) angy[ix][iz] = 0.0;
	h_angx = allocSeisHeader(nx*ny);
	h_angy = allocSeisHeader(nx*ny);
	for (ix=0;ix<nx*ny;ix++) h_angx[ix] = h_vel[ix];
	for (ix=0;ix<nx*ny;ix++) h_angx[ix].sx = sx;
	for (ix=0;ix<nx*ny;ix++) h_angy[ix] = h_vel[ix];
	for (ix=0;ix<nx*ny;ix++) h_angy[ix].sx = sx;
	compute_angles(angx,angy,wav,
			nt,ot,dt, 
			nx,ox,dx,
			ny,oy,dy,
			sx,sy,
			nz,oz,dz,sz,
			vel,fmin,fmax,
			verbose);

	if (dip_flag){
		dipx = alloc2float(nz,ntraces);
		dipy = alloc2float(nz,ntraces);
		h_dipx = allocSeisHeader(ntraces);
		h_dipy = allocSeisHeader(ntraces);
		SeisRead(dipx_name,dipx,h_dipx,&fh); 	
		SeisRead(dipy_name,dipy,h_dipy,&fh); 	
		for (imx=0;imx<nx;imx++){
			for (imy=0;imy<ny;imy++){
				for (iz=0;iz<nz;iz++) angx[imx*ny + imy][iz] += dipx[imx*ny + imy][iz];
				for (iz=0;iz<nz;iz++) angy[imx*ny + imy][iz] += dipy[imx*ny + imy][iz];
			}
		}
		free2float(dipx); 
		free2float(dipy);
		freeSeisHeader(h_dipx);
		freeSeisHeader(h_dipy);
	}

	InitFileHeader(&fh);
	fh.n1 = nz; fh.o1 = oz; fh.d1 = dz;
	fh.n2 = nx; fh.o2 = ox; fh.d2 = dx;
	fh.n3 = ny; fh.o3 = oy; fh.d3 = dy;
	SeisWrite(angx_name,angx,h_angx,&fh);
	freeSeisHeader(h_angx);
	SeisWrite(angy_name,angy,h_angy,&fh);
	freeSeisHeader(h_angy);
	free2float(angx); 
	free2float(angy); 
	free2float(vel);
	free2float(wav); 
	freeSeisHeader(h_vel);
	freeSeisHeader(h_wav);
	return 0;
}
